### DISCUSSION TITLE
samd51: MicroPython on the Seeed Wio Terminal

### DISCUSSION THREAD

**Discussion by scruss** [micropython/micropython#9838] [Show and tell]
Thanks to robert-hh's amazing efforts — made all the more amazing when you realize he hasn't got a Wio Terminal to test this on himself — MicroPython works quite well on the [Wio Terminal](https://micropython.org/download/SEEED_WIO_TERMINAL/).

In order to get folks started, I wrote a tutorial that has working examples for all operating sensors and outputs on that board: [MicroPython on the Seeed Studio Wio Terminal: it works! – We Saw a Chicken …](https://scruss.com/blog/2022/11/04/micropython-on-the-seeed-studio-wio-terminal-it-works/)

There are also shorter code snippets here: [scruss/WioTerminal-MicroPython: short MicroPython examples for the Seeed Wio Terminal](https://github.com/scruss/WioTerminal-MicroPython)

Huge thanks to Robert for making this port work so well.

**Comment by davefes** [comment]
Just read through the blurb on the website ... does the WiFi work on Micropython?

**Reply by scruss** [reply]
No; it only ever worked under Arduino. There is a small amount of hope that it might eventually, though:

1. Realtek developers have expressed interest in supporting MicroPython on and with their chips;
2. The Seeed Studio API has been changed to expose an API compatible with the ESP32 firmware used in other boards and other vendors (perhaps the same as the Nina wifi interface that's on the Arduino RP2040 Connect)

Both routes would take considerable development time, so I'm not expecting it to happen soon.

**Comment by robert-hh** [comment]
@scruss Thank you for the tutorial. It looks good and reads well. Two remarks: 
- RTC is enabled on all boards with a 32kHz crystal. So it is enabled on the WIO terminal, and can be enabled on other boards.
- The flash chip can be used for a file system using @peterhinch 's flash driver at https://github.com/peterhinch/micropython_eeprom, similar to using the SD card with the sdcard.py driver. ~~The SPI device for the flash should be 5.~~ The flash chip is connected to the QSPI interface, so access is only by SoftSPI. One topic on my list is to use a board's external flash for the default files system. That would especially be useful for the SAM21 port, which is at the limit of the flash capacity. But that would end up in a more differentiated board definition files.
- WiFi is not (yet) supported. The Seeed Wio terminal is the only board I know with WiFi support. I did not look into the capabilities of that module, but in any case it will require quite a lot of flash space. I do not know if it will fit.

Edit:
While reading through your tutorial: When I saw the connections of the Touch device I thought: Hey, that looks like a XT2046 interface. But according to the schematics it is directly connected to the MCU, and the only AF functions available at these pins are thos from the SAMD touch module, which is NOT DOCUMENTED, at least not at a level that one could write software for. If you want to integrate it, you have to use a closed software blob provided by Atmel/Microchip, which in turn requires to use their software development environment, which is the extreme opposite to their excellent data sheets. For the SAMD21, someone (adafruit?) reverse engineered the code and made a useful API for the touch module. For the SAMD51, this is still due.

**Reply by scruss** [reply]
Thanks, Robert! My Wio Terminal would have been a paperweight without your efforts. I hope your one arrives soon!

* I'm not getting any machine.RTC() on the prebuilt UF2 images. Thonny is complaining about "WARNING: Could not sync device's clock: can't import name RTC"
* Flash as filesystem would be very useful even on a SAMD51 device. The Wio Terminal has a large flash chip
* Wifi is far off. I'm not even 100% sure of the interface and API it uses. Seeed used to have an AT-compatible firmware, but they've moved on from that

As for touch, I think I got confused by the LCD_XL, LCD_YU, LCD_XR, LCD_YD pins that look terribly like the four wire output of a screen touch ADC. But there's no CS or interrupt line for the touch controller, and I'm not sure if/how it could work behind the protective case. Seeed have never advertised that the device has a touch screen, so I suspect it really doesn't.

**Reply by robert-hh** [reply]
About RTC: Right, it does not use the 32kHz clock. You could check, if this firmware boots: https://github.com/robert-hh/Shared-Stuff/blob/master/samd_firmware/SEEED_WIO_TERMINAL_firmware.uf2
This firmware also includes adc.timed() and dac.timed(), PR #9624.
Edit: Updated the file. It was a build for another board,

**Reply by robert-hh** [reply]
About the touch screen: I recall a note in a WIO terminal's document that touch will be supported in an upcoming firmware version.

**Reply by scruss** [reply]
Can confirm that RTC works as expected in that build: thanks! VBAT / PB03 / SPI_SCK is broken out to the 40-pin connector, so in theory one could apply a 3.2 V lithium battery to that for battery backup. You'd lose an external hardware SPI channel, but the internal screen / SD card / QSPI channels might be untouched.

Having some difficulty with `machine.ADC.read_timed()`, though. Doing this for the light sensor, expecting a few values returned at 50 Hz:
```
light_sensor = machine.ADC("PD01")
b=bytearray(100)
light_sensor.read_timed(b, 50)
```
I get `AttributeError: no such attribute`

**Reply by robert-hh** [reply]
Sorry, the last build did not include that. I have pushed a new version with adc.read_timed(). Too much juggling with branches. Today I made a simple python flash block device driver, which allows using the flash for a file system. Slower than the one from @peterhinch, but much smaller, which is helpful for SAMD21 devices. I have to test is a bit, but then I could add it to the default frozen bytecode set. Next would be a C version with quad flash support.

**Reply by robert-hh** [reply]
About adc.read_timed: It writes word values. So if you supply a bytearray of size 100, it puts 50 values in it.

**Reply by robert-hh** [reply]
Two days ago I mangled together a small block device driver for SPI flash, that allows using it for a file system. It is a small code, not the fastest one, and maybe just a temporary solution, but it works well with LFS1 and LFS2. This is the link https://github.com/robert-hh/SPI_Flash. When used with QSPI conncted flash on Adafruit SAMD51 or WIO Terminal, you have to use SoftSPI. The Sparkfun SAM51 Thing Plus uses a simple SPI interface for the flash.

**Reply by robert-hh** [reply]
For the WIO terminal, the start script should be:
```
import os
from machine import SoftSPI, Pin
from flashbdev import FlashBdev
wp = Pin("PA10", Pin.OUT, value=1)
hold = Pin("PA11", Pin.OUT, value=1)
cs = Pin("PB11", Pin.OUT, value=1)
spi = SoftSPI(sck=Pin("PB10"), mosi=Pin("PA08"), miso=Pin("PA09"))

flash=FlashBdev(spi, cs)
try:
    vfs = os.VfsLfs2(flash)
except OSError as e:
    print("Mount failed with error", e)
    print("Recreate the file system")
    os.VfsLfs2.mkfs(flash)
    vfs = os.VfsLfs2(flash)

os.mount(vfs, "/flash")
os.chdir("/flash")
```

**Reply by scruss** [reply]
That's really nice! I now have a 4 MB /flash filesystem that I can run main.py from

**Reply by robert-hh** [reply]
The mount script should better be:
```
import os
from machine import SoftSPI, Pin
from flashbdev import FlashBdev
wp = Pin("PA10", Pin.OUT, value=1)
hold = Pin("PA11", Pin.OUT, value=1)
cs = Pin("PB11", Pin.OUT, value=1)
spi = SoftSPI(sck=Pin("PB10"), mosi=Pin("PA08"), miso=Pin("PA09"), baudrate=1_000_000)

flash=FlashBdev(spi, cs)
try:
    vfs = os.VfsLfs1(flash, progsize=256)
except OSError as e:
    print("Mount failed with error", e)
    print("Recreate the file system")
    os.VfsLfs1.mkfs(flash, progsize=256)
    vfs = os.VfsLfs1(flash, progsize=256)

os.mount(vfs, "/flash")
```
Better use LFS1. LFS2 is pretty slow with SoftSPI, because if make gazillion small reads, when parsing through the directory structure. And if you mount the flash in boot.py, it is visible for mpremote.py.

**Reply by robert-hh** [reply]
The images in https://github.com/robert-hh/Shared-Stuff/tree/master/samd_firmware now contains a setting for maximum Baudrate with SoftSPI. You have to set baudrate to at least 1_000_000. Then, flash file access is substantially faster.

**Reply by peterhinch** [reply]
As a matter of interest did you compare the RAM consumption of your driver with mine? I'd be interested to know how big a saving you've achieved.

**Reply by robert-hh** [reply]
Actually no. Until now I just looked at the mpy file sizes. which is 1.3k vs. 3.7k. If the driver is loaded to RAM, that would bea difference. But the data RAM usage difference should be at least that of the 4k cache buffer, which is allocated by your script and makes it faster. Besides that, my driver only allocates a few bytes for local variables and the address message buffer (4-5 bytes). These size differences are important for the samd21 port with it's ~20k heap and less important for the samd51 port.

**Reply by peterhinch** [reply]
That's a substantial improvement in code size. I'm unconvinced about the buffer. Your `ioctl` returns a block size of 4096 which surely forces the littlefs driver to allocate a 4KiB RAM buffer. It seems to me that the buffer is a logical necessity because of the way Flash works.

Taking a top-down view, say you open a file in append mode, then every so often add a byte. The hardware only lets you erase 4KiB sectors. To avoid endless slow (and damaging) erases you must buffer those bytes in RAM until you have a sector full. This seems true regardless of how the device and filesystem drivers are coded. 

If I'm missing something, please do tell!

**Comment by robert-hh** [comment]
So I made a comparison, loading both drivers into RAM from the .py files and mounting the device, using `gc.collect();gc.mem_free()` before and after the load & mount.
bdevice.py + flash_spi.py: difference 12160 bytes
flashbdev.py: difference 4720 bytes
So the difference between the two is 7440 bytes or about 3k + 4k. 3k for the mpy file size difference, and 4 k for the additional cache buffer in your driver.

**Reply by peterhinch** [reply]
Thank you. That is a big improvement. 

I'm still baffled by the buffering. My code has a 4K buffer with 0.5K block size. Yours has no buffer and a 4K block size. On my reckoning the difference should be 0.5K. I'm clearly wrong.

Ff littlefs doesn't create a 4K buffer, what happens in code like this:
```python
with open("file", "a") as f:
    f.write("a")
    time.sleep(0.1)
```
The bytes must be accumulating somewhere until 4K of them have been sent. Then the chip's sector is erased and the 4K written out. Perhaps the buffer is created dynamically by extending a list? If so, this could be tested by writing 4K - 1 bytes and checking RAM.

If I'm talking nonsense please explain because I'm foxed :)

**Comment by robert-hh** [comment]
LFS1 seems to buffer data as give by the progsize parameter in the vfs constructor.
Maybe even stranger, with your flash_spi.py I have to set block_size to 12. Otherwise, LFS1 will not work properly.

**Reply by peterhinch** [reply]
Odd. I only tested LFS2.

**Reply by robert-hh** [reply]
Did you try writing files > 4kByte?

**Reply by peterhinch** [reply]
Yes. My filesystem torture test creates binary files containing random data with random lengths up to [65536 bytes](https://github.com/peterhinch/micropython_eeprom/blob/7c1496783e346572526ad024b9cd9499d3210bc1/flash/littlefs_test.py#L22).

**Reply by robert-hh** [reply]
Even with LFS2, writing large files fail unless block_size is set to 12. This is the mount script:
```
import os
from machine import SoftSPI, Pin
from flash_spi import FLASH
wp = Pin("PA10", Pin.OUT, value=1)
hold = Pin("PA11", Pin.OUT, value=1)
cspins = (Pin("PB11", Pin.OUT, value=1), )
spi = SoftSPI(sck="PB10", mosi="PA08", miso="PA09", baudrate=2_000_000)
flash=FLASH(spi, cspins, block_size=12)
try:
    vfs = os.VfsLfs2(flash, progsize=256)
except OSError as e:
    print("Mount failed with error", e)
    print("Recreate the file system")
    os.VfsLfs2.mkfs(flash)
    vfs = os.VfsLfs2(flash, progsize=256)

os.mount(vfs, "/flash")
```
Your lfs_test.py works with a few adaptions for file names and flash size. But with every flash driver I have to set the sector size to 4k, reported by ioctl.

**Reply by peterhinch** [reply]
I don't know where to take this. I tested the buffered driver extensively, against arrays of each supported flash device, running the torture test for long periods. In every test I used the driver's default littlefs block size of 512 bytes. The way the driver works is independent of the filesystem block size. This is the first report I've had of this driver requiring a different littlefs block size.

I'm afraid I'm baffled.

**Reply by robert-hh** [reply]
There is a difference between LFS1 and LFS2. 
- When with LFS2 the progsize parameter is omitted and the default of 32 is used, both lfs_test.py and my test of writing large buffers succeed, not matter which file flash driver with it's default blocksize parameter is used. When progsize is set to 256, using a blocksize of 512 fails with your driver. I did not try other value combinations. 
- When with LFS1 the progsize parameter is omitted, your driver works, but the flashbdev.py driver refuses to write files >= 4096 bytes. With a progsize of e.g. 256, both drivers work. And the blocksize returned by my driver must be 4096, the physical sector size.

**Reply by peterhinch** [reply]
Thank you, that explains my results. I never tested with LFS1 - it's not clear to me what is its use-case. Nor did I experiment with progsize. Are there any docs on the `readsize`, `progsize` and ` lookahead` args?

Unless there is a case for supporting LFS1 (nobody has requested it) my feeling is that I should just update the docs to specify supported test conditions.

**Reply by robert-hh** [reply]
The SAMD21 port only supports LFS1. Enabling LFS2 increased the build size by ~16kByte, and flash is scarce at the SAMD21. The SAMD51 variant supports LFS2 and FAT as well. LFS1 appeared to me a little bit faster in reading a directory, at the cost of worse flash utilization.
The only documentation about the readsize, progsize and lookahead args seems to be here: https://github.com/littlefs-project/littlefs/blob/master/lfs.h. More details about the implementation are here: https://github.com/littlefs-project/littlefs/blob/master/DESIGN.md
I noticed that for LFS1 the ratio blocksize/progsize must not be larger than 32, but I did not find it mentioned in the documentation. So with the 4k blocksize progsize must not be smaller than 128. Surely some implementation detail.

**Reply by peterhinch** [reply]
I think the way forward is for me to recommend LFS2 to users of my driver. For users needing a minimal flash size, I will recommend your driver with LFS1. Do you agree?

I have added a note [in this para](https://github.com/peterhinch/micropython_eeprom/blob/master/flash/FLASH.md#12-the-driver) of my docs making this observation and pointing to your solution.

**Reply by robert-hh** [reply]
Agreed. With samd21, LFS1 is anyhow the only option. If the both file systems are present, LFS2 should be used with every driver.

**Comment by ThinkTransit** [comment]
Quick question regarding the mount script, should the flash be mounted to "/" instead of "/flash"?

When mounting at "/flash" I have to install libraries in "/flash/lib" and import using

```python
from flash.lib import mylib
```

which doesn't seem correct, but I'm probably missing something!

**Reply by robert-hh** [reply]
The internal flash file system is already mounted at "/", so an additional file system has to use another path. What is the value of your sys.path?
```
import sys
print(sys.path)
```

**Reply by ThinkTransit** [reply]
```python
>>> import sys
>>> print(sys.path)
['', '.frozen']
>>> 
```
I'm using the nightly firmware from the mp website, nothing custom.

**Reply by robert-hh** [reply]
You have to add /lib to sys,path, the files in /lib are found by import:
```
import sys
sys.path.append("/lib")
```

**Reply by ThinkTransit** [reply]
Thanks @robert-hh too easy!

**Comment by ThinkTransit** [comment]
The [ili9341](https://github.com/rdagger/micropython-ili9341) micropython driver is a bit slow, would it be possible to build a c driver like https://github.com/russhughes/ili9342c_mpy into the firmware? I noticed it has been developed/tested on an esp32 but I assume if the code is c it could be also used on the samd51?

**Reply by robert-hh** [reply]
Embedding the C-Driver should be possible, but it is not likely to be accepted for the main line. Alternative approaches would be:
- Optimize the python driver.
- Use a loadable C-Code driver. That concept is rather new, and I never used it. It the interfaces to the core firmware are few, then if could be feasible.

**Reply by ThinkTransit** [reply]
Thanks @robert-hh

I'm making some progress with embedding the c driver and if I get it working I will investigate if I can make it "loadable".

**Reply by robert-hh** [reply]
Sounds good.

**Reply by ThinkTransit** [reply]
I'm getting some undefined references for lround and malloc, do you know if there are any specific headers I need to include for samd port?

```
arm-none-eabi-ld: build-SEEED_WIO_TERMINAL/src/ili9342c.o: in function `ili9342c_ILI9342C_draw':
/home/patrick/esp/ili9342c_mpy/src/ili9342c.c:493: undefined reference to `lround'
arm-none-eabi-ld: /home/patrick/esp/ili9342c_mpy/src/ili9342c.c:494: undefined reference to `lround'
arm-none-eabi-ld: /home/patrick/esp/ili9342c_mpy/src/ili9342c.c:506: undefined reference to `lround'
arm-none-eabi-ld: /home/patrick/esp/ili9342c_mpy/src/ili9342c.c:507: undefined reference to `lround'
arm-none-eabi-ld: build-SEEED_WIO_TERMINAL/src/ili9342c.o: in function `ili9342c_ILI9342C_jpg':
/home/patrick/esp/ili9342c_mpy/src/ili9342c.c:1216: undefined reference to `malloc'
arm-none-eabi-ld: /home/patrick/esp/ili9342c_mpy/src/ili9342c.c:1269: undefined reference to `free'
make: *** [Makefile:169: build-SEEED_WIO_TERMINAL/firmware.elf] Error 1

```

**Reply by ThinkTransit** [reply]
So I managed to build micropython with the c driver included however it didn't speed up the display, its still around 13 seconds for a full screen refresh.
More investigation needed!

**Reply by robert-hh** [reply]
At which baudrate did you create the SPI interface (SPI, not SoftSPI!). The maximum you can set is 48_000_000. 
What do you try to achieve, that takes 13 seconds?

**Reply by ThinkTransit** [reply]
Thanks @robert-hh the baudrate was the issue.

With a baudrate of 30_000_000 it was taking 13 seconds to clear the screen.

Reducing the baudrate down to 24_000_000 refreshes the screen in under a second.

The higher baudrate must be causing some issue somewhere.

**Reply by robert-hh** [reply]
The SPI clock is created by dividing 48Mhz by an integer. So you can have 48Mhz, 24Mhz, 16Mhz, and so on. 30_000_000 should be mapped a divisor of 1 aka 48MHz.

**Reply by russhughes** [reply]
I see the same behavior. Hardware SPI speeds higher than half of the SERCOM clock are extremely slow.

**Reply by robert-hh** [reply]
I'll check if I can find a reason for that. Otherwise I will limit the baudrate to SERCOM_clock/2.

**Reply by robert-hh** [reply]
Ok. So it's SERCOM_CLOCK/(2 * baudrate) - 1. So using 48Mhz results in the largest possible divider. I prepared a fix avoiding the underflow. Will come with the next service pack.

**Reply by robert-hh** [reply]
@ThinkTransit Did you check the Python ILI9341 driver with the other baudrate as well for speed? And there is as well a much simpler IL9341 driver using the framebuffer by @peterhinch https://github.com/peterhinch/micropython-nano-gui/tree/master/drivers/ili93xx

**Reply by russhughes** [reply]
I saw the same results using the driver https://github.com/rdagger/micropython-ili9341. While https://microchipsupport.force.com/s/article/SPI-max-clock-frequency-in-SAMD-SAMR-devices specifically references the samd21 I suspect this limit applies to the samd51 as well.

**Reply by ThinkTransit** [reply]
@robert-hh Yes I tested Peter's python driver @24_000_000 and it works fine.

I have micro-gui running on the WIO now which is pretty cool. The only issue I have is that it complains about complex numbers not being supported. Do you know if this is an architecture limitation or is there work required to get complex numbers working?

https://www.youtube.com/watch?v=3nsCb7HmfaE&ab_channel=DunnyRoll

**Reply by robert-hh** [reply]
Complex numbers are simply not enabled in the config options. See mcu/samd51/mpconfigmcu.h, where the options are set to 0 to make it obvious. Can be enabled.

**Reply by ThinkTransit** [reply]
Thanks @robert-hh that worked!

**Comment by ThinkTransit** [comment]
@scruss great article

One trap for new players is that if the directory "SD" doesn't exist then the SD card will not mount and report "SD card not found", could be good to add a note in the article to create "SD" or catch the OSError: [Errno 1] EPERM error.

**Reply by scruss** [reply]
I have never needed to create the directory: `os.mount(sd, "/SD")` appears to do it for me. Are you sure this is necessary? It is under Linux, but there's no OS to worry about on a SAMD51 board

**Reply by robert-hh** [reply]
os.mount() creates the entry for the mount point.

**Reply by ThinkTransit** [reply]
Apologies @scruss It must have been a coincidence that it failed and then worked after I created the directory!

**Comment by robert-hh** [comment]
As an exercise, I made a C version of the raw flaw driver class in flashbdev.py. It's a little bit smaller than the .py version, and when operating with a hard SPI interface, file accesses are more snappy, especially the short ones. With SoftSPI, there is no noticeable difference. As next step I plan to add a QSPI flash driver, which gives hardware speed to the flash of SAMD51 devices like the WIO terminal or Adafruit SAMD51 boards. The branch is here: https://github.com/robert-hh/micropython/tree/samd_spiflash

**Comment by robert-hh** [comment]
The branch at https://github.com/robert-hh/micropython/tree/samd_spiflash is now extended by a C-Code QSPI flash class in machine. It is based on a Adafruit QSPI driver. That speeds up flash access on SAMD51 devices with QSPI Flash substantially. It uses the same small flashbdev.py driver as the SPI flash class. If you build the firmware yourself, this driver is included in the frozen bytecode. Otherwise, it's below.
```
SECTOR_SIZE = const(4096)


class FlashBdev:
    def __init__(self, flash, sectorsize=SECTOR_SIZE):
        self.flash = flash
        self.sectorsize = sectorsize

    def readblocks(self, n, buf, offset=0):
        self.flash.readblock(n * self.sectorsize + offset, buf)

    def writeblocks(self, n, buf, offset=0):
        if offset == 0:
            self.flash.erase(n * self.sectorsize)
        self.flash.writeblock(n * self.sectorsize + offset, buf)

    def ioctl(self, op, arg):
        if op == 4:  # MP_BLOCKDEV_IOCTL_BLOCK_COUNT
            return self.flash.getsize() // self.sectorsize
        if op == 5:  # MP_BLOCKDEV_IOCTL_BLOCK_SIZE
            return self.sectorsize
        if op == 6:  # MP_BLOCKDEV_IOCTL_BLOCK_ERASE
            self.flash.erase(arg * self.sectorsize)
            return 0
```
When I was almost finished, I found Damiens spiflash.c driver in drivers/memory. It is used by the STM32 port. Not sure how to use that. It would need an interface for the hardware and to the Python API. The new SAMD drivers mostly of these interface parts, and the code in /drivers/memory would add a large layer in between.

**Comment by robert-hh** [comment]
I renamed the samd_spiflash branch into samd_flash. The commits are rearranged to have fewer of them, and the last commit changes the startup code _boot.py to use the board's flash drive as default drive. With SAMD51 QSPIflash, that increase the CPU speed as well. My test code pystone_lowmem.py runs about two times faster. With SAM21 devices there is no speed gain, but at least a larger file system. I moved the additional flash drivers from machine into samd. It still may be optimized. Having three different flash drivers does not look nice.
